### PR TITLE
Add specific handling for variable aliasing in To_cmm

### PR DESCRIPTION
**Pull request by Gbury**
First part split from #1087 .

This add specific handling for variable aliasing.

Consider the following standalone example extracted from `bigstring` :

```ocaml
let arch_big_endian = Sys.big_endian
let not_on_32bit = Sys.word_size > 32
let mask32_n = Nativeint.(sub (shift_left 1n 32) 1n)

type t =
    ( char
    , Stdlib.Bigarray.int8_unsigned_elt
    , Stdlib.Bigarray.c_layout )
      Stdlib.Bigarray.Array1.t

external swap32 : int32 -> int32 = "%bswap_int32"
external int32_to_int : int32 -> int = "%int32_to_int"
external unsafe_get_32 : t -> int -> int32 = "%caml_bigstring_get32u"

let[@inline always] uint32_of_int32_t n =
  if not_on_32bit
  then
    (* use Caml.Nativeint to ensure inlining even without x-library-inlining *)
    Nativeint.(to_int (logand (of_int32 n) mask32_n))
  else int32_to_int n
;;

let unsafe_read_int32 t ~pos = unsafe_get_32 t pos
let unsafe_read_int32_swap t ~pos = swap32 (unsafe_get_32 t pos)

let unsafe_get_int32_t_be =
  if arch_big_endian then unsafe_read_int32 else unsafe_read_int32_swap
;;

let[@inline] unsafe_get_uint32_be t ~pos =
  uint32_of_int32_t (unsafe_get_int32_t_be t ~pos)
;;
```

We now generate the following cmm code (when in `-Oclassic` mode):
```
(function{/home/guigui/tmp/bigstring.ml:31,34-93}
 camlBigstring__unsafe_get_uint32_be_3_3_code
     (t106/588: val pos107/589: val my_closure105/590: val)
 (let
   (to_cmm_split_12/597
      (let ba_data/594 (load_mut int (+a t106/588 8))
        (load_mut signed int32 (+ ba_data/594 (>>s pos107/589 1))))
    to_cmm_split_16/601
      (>>s (<< (bswap_32 (>>s (<< to_cmm_split_12/597 32) 32)) 32) 32)
    region123/606 (beginregion))
   (catch
     (if (== 1 (load val (+a (load val (+a my_closure105/590 24)) 16)))
       (exit 2 (or (>>s (<< to_cmm_split_16/601 32) 31) 1))
       (let
         Pandbint135/620
           (alloc_local{/home/guigui/tmp/bigstring.ml:20,22-52} 2815
             "caml_nativeint_ops"
             (and (>>s (<< to_cmm_split_16/601 32) 32)
               (load int
                 (+a (load val (+a (load val (+a my_closure105/590 24)) 24))
                   8))))
         (exit 2 (+ (<< (load int (+a Pandbint135/620 8)) 1) 1))))
   with(2 region_return124/607: val)
     (let
       (pos110/592 pos107/589 t111/593 t106/588
        n118/602
          (alloc{/home/guigui/tmp/bigstring.ml:25,36-64} 2303
            "caml_int32_ops" (>>s (<< to_cmm_split_16/601 32) 32))
        my_closure120/603 (load val (+a my_closure105/590 24))
        to_cmm_split_20/609 (load val (+a my_closure105/590 24))
        not_on_32bit121/604 (load val (+a to_cmm_split_20/609 16))
        to_cmm_split_22/608 (load val (+a my_closure105/590 24))
        mask32_n122/605 (load val (+a to_cmm_split_22/608 24))
        unit125/610 (seq (endregion region123/606) 1))
       region_return124/607))))
```

whereas before (i.e. with `main`) we would generate the following (again, when in `-Oclassic` mode):
```(function{/home/guigui/tmp/bigstring.ml:31,34-93}
 camlBigstring__unsafe_get_uint32_be_3_3_code
     (t106/588: val pos107/589: val my_closure105/590: val)
 (let
   (to_cmm_split_12/597
      (let ba_data/594 (load_mut int (+a t106/588 8))
        (load_mut signed int32 (+ ba_data/594 (>>s pos107/589 1))))
    n118/602
      (alloc{/home/guigui/tmp/bigstring.ml:25,36-64} 2303 "caml_int32_ops"
        (>>s (<< (bswap_32 (>>s (<< to_cmm_split_12/597 32) 32)) 32) 32))
    region122/606 (beginregion))
   (catch
     (if (== 1 (load val (+a (load val (+a my_closure105/590 24)) 16)))
       (exit 2 (+ (<< (load signed int32 (+a n118/602 8)) 1) 1))
       (let
         Pandbint134/620
           (alloc_local{/home/guigui/tmp/bigstring.ml:20,22-52} 2815
             "caml_nativeint_ops"
             (and (load signed int32 (+a n118/602 8))
               (load int
                 (+a (load val (+a (load val (+a my_closure105/590 24)) 24))
                   8))))
         (exit 2 (+ (<< (load int (+a Pandbint134/620 8)) 1) 1))))
   with(2 region_return123/607: val)
     (let
       (pos110/592 pos107/589 t111/593 t106/588
        my_closure119/603 (load val (+a my_closure105/590 24))
        to_cmm_split_21/609 (load val (+a my_closure105/590 24))
        not_on_32bit120/604 (load val (+a to_cmm_split_21/609 16))
        to_cmm_split_23/608 (load val (+a my_closure105/590 24))
        mask32_n121/605 (load val (+a to_cmm_split_23/608 24))
        unit124/610 (seq (endregion region122/606) 1))
       region_return123/607))))
```

The main difference is that with this PR, we now currently push down the `Int32` allocation of `n118` down at its use site inside the `catch`. There is still an unneeded flush of the allocation before the `end_region` at the end of the function, but that one will be removed in the upcoming second part of the split of #1087 that computes free_names during `to_cmm`.

### CODE REVIEW

**Comment by mshinwell** [comment]
@lthls will review this (please ping @mshinwell before merging)

File: middle_end/flambda2/to_cmm/to_cmm_env.ml [hunk 1]
```diff
@@ -783,6 +794,72 @@ let inline_variable ?consider_inlining_effectful_expressions env res var =
...
+  let env = { env with inline_once_aliases } in
+  env, res
+
+(* Situation: [alias_of] is a must_inline (once or duplicate), and [var] is used
+   more than once.
+
+   In this case, we want to force splitting of the original binding, and then
+   bind the new variable with a `must_inline` inline status *)
+let split_binding_and_rebind ~num_occurrences_of_var env res var alias_of
+    binding =
```
Comment (lthls) [inline middle_end/flambda2/to_cmm/to_cmm_env.ml:822]: Some of those parameters could likely use labels.

File: middle_end/flambda2/to_cmm/to_cmm_env.ml [hunk 2]
```diff
@@ -589,12 +597,14 @@ and split_in_env env res var binding =
...
-      | Must_inline_once -> env
-      | Must_inline_and_duplicate ->
+      let[@inline] update_env () =
         { env with
           bindings = Variable.Map.add var (Binding split_binding) env.bindings
         }
+      in
+      match split_binding.inline with
+      | Must_inline_once -> env
+      | Must_inline_and_duplicate -> update_env ()
```
[the remark this answers is not in the document]
Comment (Gbury) [reply middle_end/flambda2/to_cmm/to_cmm_env.ml]: I removed the change

File: middle_end/flambda2/to_cmm/to_cmm_env.ml (hunk 1 above)
Comment (Gbury) [reply middle_end/flambda2/to_cmm/to_cmm_env.ml:820]: I added labels to `~var` and `~alias_of`

**Comment by lthls** [comment]
@mshinwell We're finished with this PR, we can merge whenever you're ready.

**Pull request merged**